### PR TITLE
Adding an optional `name` parameter in Parameter.serialize in order to make warnings more helpful

### PR DESCRIPTION
**Pull request by Lucas-C**
## Description
I added an optional `name` Parameter in https://github.com/voyages-sncf-technologies/luigi/blob/master/luigi/parameter.py#L247 in order to improve warnings.
I also passed this new parameter in many place where `Parameter.serialize` is invoked (not all, only when possible).
Finally I updated existing tests and added new ones.

## Motivation and Context
When this `name` parameter is passed to `Parameter.serialize`, warnings raised will now include the parameter name, helping to diagnose the issue.
This is especially helpful in https://github.com/spotify/luigi/blob/master/luigi/task.py#L473 this is called to validate `luigi.cfg`.

## Have you tested this? If so, how?
I have included unit tests and I used the patched version in our own luigi pipelines.

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -155,6 +155,7 @@ Some more companies are using Luigi but haven't had a chance yet to write about
 * `OAO <https://adops.com/>`_
 * `Grovo <https://grovo.com/>`_
 * `Weebly <https://www.weebly.com/>`_
+* `Voyages-Sncf <https://voyages-sncf.com>`_
```
Comment (dlstadther) [inline README.rst]: Could you separate this into a independent PR? Thanks

Comment (Lucas-C) [inline README.rst]: Sure. Done

File: luigi/parameter.py [hunk 2]
```diff
@@ -244,16 +244,17 @@ def _parse_list(self, xs):
...
         Opposite of :py:meth:`parse`.
 
         Converts the value ``x`` to a string.
 
         :param x: the value to serialize.
+        :param name: the parameter name, for reporting purposes in case of an error.
         """
         if not isinstance(x, six.string_types) and self.__class__ == Parameter:
-            warnings.warn("Parameter {0} is not of type string.".format(str(x)))
+            warnings.warn("Parameter {0} is not of type string.".format(name or x))
```
Comment (Tarrasch) [inline luigi/parameter.py]: Is this the only place where you use `name`?

**Comment by Tarrasch** [comment]
Just to clarify, I added a different way of solving this, because I think this is a great idea, and I could have used it myself many times. But I think this implementation could be better, and that in the long run it saves time for everyone: Users, you (the contributor) and me (the maintainer).

File: luigi/parameter.py (hunk 2 above)
Comment (Tarrasch) [inline luigi/parameter.py]: also, was removing `str()` on purpose? :)

Comment (Lucas-C) [inline luigi/parameter.py]: In this function yes. And also in the overriden version in TimeDeltaParameter.serialize

Comment (Lucas-C) [inline luigi/parameter.py]: Yes, the removal was on purpose, I think it was unneeded

**Comment by Lucas-C** [comment]
Alright, this is a good suggestion as it will reduce the impact of this change.
However IHMO adding a new function slightly adds in complexity.
I'll update this PR tomorrow.

**Comment by Lucas-C** [comment]
New version pushed. It is definitively cleaner I think.

**Comment by Lucas-C** [comment]
Tested with:
```
tox -e py34-nonhdfs --run-command "python test/runtests.py -v test/task_test.py"
```

File: test/task_test.py [hunk 3]
```diff
@@ -106,6 +109,28 @@ def test_no_unpicklable_properties(self):
         message = task.set_status_message('message')
         self.assertEqual(message, 'message')
 
+    def test_no_warn_if_param_types_ok(self):
+        with warnings.catch_warnings(record=True) as w:
+            DummyTask(**DUMMY_TASK_OK_PARAMS)
+        self.assertEqual(len(w), 0, msg='No warning should be raised when correct parameter types are used')
+
+    if six.PY3:  # assertWarnsRegex was introduced in Python 3.2
```
Comment (Tarrasch) [inline test/task_test.py]: Nice :)

File: test/task_test.py [hunk 4]
```diff
@@ -106,6 +109,28 @@ def test_no_unpicklable_properties(self):
...
+        def test_warn_on_non_str_param(self):
+            params = dict(**DUMMY_TASK_OK_PARAMS)
```
Comment (Tarrasch) [inline test/task_test.py]: Can't you just use `DUMMY_TASK_OK_PARAMS` directly? :)

Comment (Lucas-C) [inline test/task_test.py]: This test checks that an integer value (42) given to the "param" parameter (that expects a string) triggers a warning.
Hence I need to modify a copy of `DUMMY_TASK_OK_PARAMS`.

**Comment by Tarrasch** [comment]
Can you fix the flake issue please? Then I think we can merge. :)

**Comment by Lucas-C** [comment]
Done
